Add lesson 2 (#6)
EugeniyBurmistrov/resolve: docs/basics/tutorial/lesson-2.md

## **Lesson 2** - Write side - Add a List Item

[\[Get the Code for This Lesson\]](https://github.com/reimagined/resolve/tree/master/examples/shopping-list-tutorial/lesson-2)

> This lesson will teach you how to implement a basic write side for your reSolve application. An application's [write side](resolve-app-structure.md#write-and-read-sides) handles commands, performs input validation and emits **events** based on valid commands. The framework then saves the emitted events to the **event store**.

Reviewer: ```suggestion
This lesson describes how to implement a basic write side for a reSolve application. An application's [write side](resolve-app-structure.md#write-and-read-sides) handles commands, performs input validation, and emits **events** based on valid commands. The framework then saves the emitted events to the **event store**.
```

> In the CQRS and Event Sourcing paradigms, commands are handled by Domain Objects grouped into aggregates. ReSolve implements aggregates as static objects that contain sets of functions. These functions can be of one of the following two kinds:

Reviewer: ```suggestion
In the CQRS and Event Sourcing paradigms, Domain Objects grouped into aggregates handle commands. ReSolve implements aggregates as static objects that contain sets of functions. These functions can be of one of the following types:
```

> - **[Command Handlers](write-side.md#aggregate-command-handlers)** - Handle commands and emit events in response.

Reviewer: emit?

Author: Articles about the Event Sourcing paradigm mostly say that aggregates 'emit' events. 
I considered to use the word 'generate', but it would not be accurate for every case. Technically, an event can be generated by some external service or helper and then emitted (or published) by an aggregate.

Reviewer: raise?

Author: I need to point out that here we talk about an event not as an 'occurence in a point of time'. In the Event Sourcing paradigm, an event is an abstract entity that we can take and save to the storage.

Technical sources say 'emit' or 'publish'.

Reviewer: This is the first time I've seen anyone use 'emit' like this, that is why I'm questioning its use.

> - **[Projections](write-side.md#aggregate-projection-function)** - Build aggregate state from events so this state can be observed on the write side, for example to perform input validation.

Reviewer: Why "observed"?

Author: I am trying to say that your write side implementation can check what is currently in the state to make some arbitrary decisions. Maybe the word "inspected" would be more correct.

Reviewer: so this state can be checked?

Author: Fair enough, changed to 'checked'

### Creating an Aggregate

> Use the following steps to implement the write side for your shopping list application.

Reviewer: ```suggestion
Use the following steps to implement the write side of your shopping list application:
```

> To add an aggregate to you shopping list application, first define types of events that this aggregate can produce. Create an **eventTypes.js** file in the project's **common** folder and add the following content to it.

Reviewer: ```suggestion
To add an aggregate to your shopping list application, define types of events that this aggregate can produce. Create an **eventTypes.js** file in the project's **common** folder and add the following content to it:
```

**common/eventTypes.js:**

<!-- prettier-ignore-start -->

[embedmd]:# (../examples/shopping-list-tutorial/lesson-2/common/eventTypes.js /^/ /\n$/)
```js
export const SHOPPING_LIST_CREATED = "SHOPPING_LIST_CREATED";

export const SHOPPING_ITEM_CREATED = "SHOPPING_ITEM_CREATED";
```

<!-- prettier-ignore-end -->

> For now, your application requires only two event types:

Reviewer: For now?

Author: Wrote so because a reader needs to add more event types in further lessons as their application becomes more complex.

Reviewer: If the reader needs to add something later, you can just write these things later.

Reviewer: If they only need two now, simply state they should add these now and the reason why.

Author: Removed this part and moved descriptions to comments in the code sample:

```js
export const SHOPPING_LIST_CREATED = 'SHOPPING_LIST_CREATED' // Signals about creation of a shopping list

export const SHOPPING_ITEM_CREATED = 'SHOPPING_ITEM_CREATED' // Signals about creation of an item within a shopping list
```

Reviewer: Why 'signals'?

Author: 'indicates the creation'?

- SHOPPING_LIST_CREATED - Signals about creation of a shopping list;
- The SHOPPING_ITEM_CREATED - Signals about creation of an item within a shopping list.

> Next, create a **shopping_list.commands.js** file in the **common/aggregates** folder. This file will contain command handlers for the ShoppingList aggregate. Add the following code to the file:

Reviewer: ```suggestion
Next, create a **shopping_list.commands.js** file in the **common/aggregates** folder to store command handlers for the ShoppingList aggregate. Add the following code to the file:
```

**common/aggregates/shopping_list.commands.js:**

```js
import { SHOPPING_LIST_CREATED, SHOPPING_ITEM_CREATED } from '../eventTypes'

export default {
  createShoppingList: (state, { payload: { name } }) => {
    return {
      type: SHOPPING_LIST_CREATED,
      payload: { name }
    }
  },
  createShoppingItem: (state, { payload: { id, text } }) => {
    return {
      type: SHOPPING_ITEM_CREATED,
      payload: { id, text }
    }
  }
}
```

> This file exports an object with two command handlers. A command handler receives the aggregate state and a command payload. A payload can contain any arbitrary data related to the command. For example, the **createShoppingList** command's payload contains a shopping list name, and the **createShoppingItem** command payload contains an item's ID and text to display.

Reviewer: ```suggestion
This file exports an object with two command handlers. A command handler receives the aggregate state and a command payload. A payload can contain any arbitrary data related to the command. For example, the **createShoppingList** command's payload contains the shopping list's name, and the **createShoppingItem** command payload contains an item's ID and display text.
```

> A command handler returns an event object. This object should contain the following obligatory fields:

Reviewer: ```suggestion
A command handler returns an event object. This object should contain the following fields:
```

Reviewer: Why do you add 'obligatory'?

- **type** - specifies the event's type;
- **payload** - specifies data associated with the event.

> In the example code, the event payload contains the same fields that were obtained from the command payloads. The reSolve framework saves events returned by command handlers to a persistent **[event store](write-side.md#event-store)**. For now, your application is configured to use a file-based event store, which is sufficient for learning purposes.

Reviewer: "events returned by command handlers" Why do you use passive voice here?

Reviewer: and again, why 'For now'?

Reviewer: "sufficient for learning purposes"?

Author: > and again, why 'For now'?

See my comment above

Author: > "events returned by command handlers" Why do you use passive voice here?

"The reSolve framework saves events that command handlers return to a persistent **[event store](write-side.md#event-store)**."

Author: > "sufficient for learning purposes"?

I'm trying to say that we use a primitive configuration because this is a tutorial and we need to start from the basics. 

"sufficient for a tutorial"?

As a more detailed alternative:

"Your application is already configured to use a file-based event store. We suggest that you keep this configuration throughout the tutorial. For information on how to use other storage types see the following documentation topics:"
- [Adapters](https://github.com/reimagined/resolve/blob/master/docs/advanced-techniques.md#adapters)
- [Configuring Adapters](https://github.com/reimagined/resolve/blob/master/docs/preparing-to-production.md#configuring-adapters)

> Your minimal shopping list aggregate is now ready. The last step is to register it in the application's configuration file. Open the **config.app.js** file, locate the **aggregates** configuration section and specify the following settings:

Reviewer: Why 'minimal'?

Reviewer: ```suggestion
Your shopping list aggregate is now ready. The last step is to register it in the application's configuration file. To do this, open the **config.app.js** file and specify the following settings in the **aggregates** configuration section:
```

**config.app.js:**

```js
...
aggregates: [
  {
    name: 'ShoppingList',
    commands: 'common/aggregates/shopping_list.commands.js',
  }
],
...
```

### Sending Commands to an Aggregate

> Now that your application can commands, you can use the reSolve framework's standard HTTP API to send such commands to create a shopping list and populate it with items.

Reviewer: "can commands"??

Author: Missed a word. 
"can handle commands"

A request body should have the `application/json` content type and contain a JSON representation of the command:

```
{
  "aggregateName": "ShoppingList",
  "type": "createShoppingList",
  "aggregateId": "shopping-list-1",
  "payload": {
    "text": "Item 1"
  }
}
```

> In addition to the aggregate name, command type and payload, this object specifies the aggregate Id (a unique identifier of an aggregate instance).

Reviewer: ```suggestion
In addition to the aggregate name, command type and payload, this object specifies the aggregate's ID (the aggregate instance's unique identifier).
```

Reviewer: Although, I'm not sure it is necessary to explain what the ID is.

Author: You misinterpreted this part. *Aggregate ID* is a term. Here
'ID' refers not to the aggregate (abstract) but to an aggregate instance (concrete). We added the explanation to clarify this exact thing.

Reviewer: How can 'Aggregate ID' be a term if one part of it refers to something else? That is like saying "He lives at 32 Thornton Lane (the 32 is his age)"? If you write 'Aggregate ID', it means aggregate ID.

Author: What I'm saying is that there are two things that you can call 'aggregate' - a definition in code and an instance created from this definition at runtime (similar to classes and objects in OOP).  But I see the problem. Changed to your version but without the explanation in parentheses.

Run your application and send a POST request to the following URL:

```
http://127.0.0.1:3000/api/commands
```

> You can do this using any REST client or using **curl**. For example, use the following inputs to create a shopping list:

Reviewer: using... using?

Author: 'You can use any REST client or **curl** to do this.'

--header "Content-Type: application/json" \
--data '
{
    "aggregateName": "ShoppingList",
    "aggregateId": "shopping-list-1",
    "type": "createShoppingList",
    "payload": {
        "name": "List 1"
    }
}
'

X-Powered-By: Express
Content-Type: text/plain; charset=utf-8
Date: Wed, 19 Dec 2018 12:16:56 GMT
Connection: keep-alive
Content-Length: 139

{"type":"SHOPPING_LIST_CREATED","payload":{"name":"List 1"},"aggregateId":"shopping-list-1","aggregateVersion":1,"timestamp":1545221816663}

```

Use the inputs shown below to add an item to the created shopping list:

```sh
curl -i http://localhost:3000/api/commands/ \
--header "Content-Type: application/json" \
--data '
{
    "aggregateName": "ShoppingList",
    "aggregateId": "shopping-list-1",
    "type": "createShoppingItem",
    "payload": {
        "id": "1",
        "text": "Milk"
    }
}
'

X-Powered-By: Express
Content-Type: text/plain; charset=utf-8
Date: Wed, 19 Dec 2018 12:17:53 GMT
Connection: keep-alive
Content-Length: 146

{"type":"SHOPPING_ITEM_CREATED","payload":{"id":"1","text":"Milk"},"aggregateId":"shopping-list-1","aggregateVersion":2,"timestamp":1545221873644}

```

> Add a few more more items to have data to work with in future lessons.

Reviewer: Why do you add this sentence?

Author: Removed the sentence

Reviewer: Ok

> Now, you can check the event store file to see the newly created event. Open the **event-storage.db** file and locate the created event objects:

Reviewer: ```suggestion
You can now check the event store file to see the newly created event. Open the **event-storage.db** file and locate the created event objects:
```

<!-- prettier-ignore-start -->

``` json
{"type":"SHOPPING_LIST_CREATED","payload":{"name":"List 1"},"aggregateId":"shopping-list-1","aggregateVersion":1,"timestamp":1542884752421,"aggregateIdAndVersion":"shopping-list-1:1","_id":"Ujiz4pjVwid1AaZP"}
{"type":"SHOPPING_ITEM_CREATED","payload":{"id":"1","text":"Milk"},"aggregateId":"shopping-list-1","aggregateVersion":2,"timestamp":1542884835201,"aggregateIdAndVersion":"shopping-list-1:2","_id":"RBr1596unUVhTJeo"}
{"type":"SHOPPING_ITEM_CREATED","payload":{"id":"2","text":"Eggs"},"aggregateId":"shopping-list-1","aggregateVersion":3,"timestamp":1542884852629,"aggregateIdAndVersion":"shopping-list-1:3","_id":"WJfG65khmyoPY12U"}
{"type":"SHOPPING_ITEM_CREATED","payload":{"id":"3","text":"Canned beans"},"aggregateId":"shopping-list-1","aggregateVersion":4,"timestamp":1542884875144,"aggregateIdAndVersion":"shopping-list-1:4","_id":"qvKCvnEOhVQrD7xJ"}
{"type":"SHOPPING_ITEM_CREATED","payload":{"id":"4","text":"Paper towels"},"aggregateId":"shopping-list-1","aggregateVersion":5,"timestamp":1542884890484,"aggregateIdAndVersion":"shopping-list-1:5","_id":"YEnzkAlBjEqaLwQI"}

```

<!-- prettier-ignore-end -->

### Performing Validation

> Your application's write side currently does not perform any input validation. This results in the following flaws:

Reviewer: "the following flaws"? flaws?

Author: 'issues'?

Reviewer: Ok

> - The command handlers do not check whether or not all required fields are provided in a command's payload.

Reviewer: Why do you use 'or not'?

Author: We are used to write 'whether or not' because in the past Armina insisted that we use this form. I will just write 'whether' from now on.

- It is possible to create more then one shopping list with the same aggregate ID.

> - You can create items in a shopping list that does not exist.

Reviewer: items... does?

Author: I mean, a shopping list does not exist

Author: 'in a nonexistent list'?

> You can overcome the first flaw by adding simple checks to each command handler:

Reviewer: Why 'flaw'?

**common/aggregates/shopping_list.commands.js:**

```js
createShoppingList: (state, { payload: { name } }) => {
  if (!name) throw new Error("name is required");
  ...
},
createShoppingItem: (state, { payload: { id, text } }) => {
  if (!id) throw new Error('id is required')
  if (!text) throw new Error('text is required')
  ...
}
```

To overcome the second and third flaws, you can use the **aggregate state**. This state is assembled on the fly by an aggregate **projection** from previously created events. To add a projection to the ShoppingList aggregate, create a **shopping_list.projection.js** file in the **common/aggregates** folder and add the following code there:

**common/aggregates/shopping_list.projection.js:**

<!-- prettier-ignore-start -->

[embedmd]:# (../examples/shopping-list-tutorial/lesson-2/common/aggregates/shopping_list.projection.js /^/ /\n$/)
```js
import { SHOPPING_LIST_CREATED } from "../eventTypes";

export default {
  Init: () => ({}),
  [SHOPPING_LIST_CREATED]: (state, { timestamp }) => ({
    ...state,
    createdAt: timestamp
  })
};
```

<!-- prettier-ignore-end -->

> Register the create projection in the application configuration file:

Reviewer: ```suggestion
Register the create projection in the application's configuration file:
```

**config.app.js:**

<!-- prettier-ignore-start -->

[embedmd]:# (../examples/shopping-list-tutorial/lesson-2/config.app.js /^[[:blank:]]+aggregates:/ /\],/)
```js
  aggregates: [
    {
      name: "ShoppingList",
      commands: "common/aggregates/shopping_list.commands.js",
      projection: "common/aggregates/shopping_list.projection.js"
    }
  ],
```

<!-- prettier-ignore-end -->

> The projection object should specify an obligatory **Init** function and a set of **projection functions**.

Reviewer: "should specify an obligatory" Why do you use 'should' and 'obligatory'?

Author: The projection object specifies an **Init** function and a set of **projection functions**.

- The Init function initializes the aggregate state. In the example code, it creates a new empty object.
- Projection functions build the aggregate state based on the aggregate's events. Each such function is associated with a particular event type. The function receives the previous state and an event, and returns a new state based on the input.

> In the example code, the SHOPPING_LIST_CREATED projection function adds the SHOPPING_LIST_CREATED event's timestamp to the state. This information can be used on the write side to find out whether and when a shopping list has been created for the current aggregate instance (i.e., an instance identified by the current aggregate ID).

Reviewer: Why 'has been created'? This is something that happened in the past, so why do you not use past simple?

Reviewer: I've told you this many times, do not use 'i.e.' and do not use passive voice when you can use active voice: identified by the current aggregate ID -> the current aggregate ID identifies

Author: ok

**common/aggregates/shopping_list.commands.js:**

```js
  createShoppingList: (state, { payload: { name } }) => {
    if (state.createdAt) throw new Error("shopping List already exists");
    ...
  },
  createShoppingItem: (state, { payload: { id, text } }) => {
    if (!state || !state.createdAt) {
      throw new Error(`shopping list does not exist`);
    }
    ...
  }
```

> You can check whether or not the validation works as intended by sending commands to your aggregate:

Reviewer: whether or not?

Author: Changed to 'whether'

```sh
# Trying to create a shopping list without specifying the name
$ curl -i http://localhost:3000/api/commands/ \
> --header "Content-Type: application/json" \
> --data '
> {
>     "aggregateName": "ShoppingList",
>     "aggregateId": "shopping-list-2",
>     "type": "createShoppingList",
>     "payload": { }
> }
> '
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   164  100    31  100   133    142    610 --:--:-- --:--:-- --:--:--   655HTTP/1.1 500 Internal Server Error
X-Powered-By: Express
Date: Thu, 22 Nov 2018 11:14:10 GMT
Connection: keep-alive
Content-Length: 31

Command error: name is required

> # Trying to create a shopping list that already exists

Reviewer: ```suggestion
# When you create a shopping list that already exists
```

$ curl -i http://localhost:3000/api/commands/ \
> --header "Content-Type: application/json" \
> --data '
> {
>     "aggregateName": "ShoppingList",
>     "aggregateId": "shopping-list-1",
>     "type": "createShoppingList",
>     "payload": {
>         "name": "List 1"
>     }
> }
> '
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   205  100    43  100   162    196    739 --:--:-- --:--:-- --:--:--   798HTTP/1.1 500 Internal Server Error
X-Powered-By: Express
Date: Thu, 22 Nov 2018 11:11:18 GMT
Connection: keep-alive
Content-Length: 43

> Command error: shopping list already exists

Reviewer: ```suggestion
Command error: the shopping list already exists
```

# Trying to add an item to an inexistent shopping list
$ curl -i http://localhost:3000/api/commands/ \
> --header "Content-Type: application/json" \
> --data '
> {
>     "aggregateName": "ShoppingList",
>     "aggregateId": "shopping-list-4000",
>     "type": "createShoppingItem",
>     "payload": {
>         "id": "5",
>         "text": "Bread"
>     }
> }
> '
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   226  100    43  100   183    211    901 --:--:-- --:--:-- --:--:--   901HTTP/1.1 500 Internal Server Error
X-Powered-By: Express
Date: Thu, 22 Nov 2018 11:16:56 GMT
Connection: keep-alive
Content-Length: 43

> Command error: shopping list does not exist

Reviewer: ```suggestion
Command error: the shopping list does not exist
```
